docs/llvm/pass/hello-world.md

```diff
@@ -0,0 +1,135 @@
+
+In this section, we guide you through writing and running a simple pass that prints the names of all functions contained in the input IR module.
+
+## Setup the Build
+
+Let's assume that you have cloned the [LLVM monorepo] and setup the build directory at `build/` [^how2build], and the current working directory is the root of the monorepo. In this section, we will build the new pass in-tree, which means that the source tree of the new pass is directly embedded in the source tree of the LLVM monorepo.
+
+[LLVM monorepo]: https://github.com/llvm/llvm-project
+[^how2build]: For instructions on how to build LLVM from source, please refer to the [official documentation](https://llvm.org/docs/GettingStarted.html#getting-the-source-code-and-building-llvm).
+
+Create a `Hello.h` file under `llvm/include/llvm/Transforms/Utils/`:
+
+```sh
+touch llvm/include/llvm/Transforms/Utils/Hello.h
+```
+
+Create a `Hello.cpp` file under `llvm/lib/Transforms/Utils/`:
+
+```sh
+touch llvm/lib/Transforms/Utils/Hello.cpp
+```
+
+Modify `llvm/lib/Transforms/Utils/CMakeLists.txt` and add `Hello.cpp` to the `add_llvm_component_library` call:
+
+```CMake
+add_llvm_component_library(LLVMTransformUtils
+    # Other files ...
+    Hello.cpp
+    # Other files and configurations ...
+)
+```
+
+This step adds `Hello.cpp` to the LLVM build process.
+
+## Code Implementation
+
+Edit `Hello.h` as follows:
+
+```cpp
+#ifndef LLVM_TRANSFORMS_HELLO_H
+#define LLVM_TRANSFORMS_HELLO_H
+
+#include "llvm/IR/PassManager.h"
+
+namespace llvm {
+
+class HelloPass : public PassInfoMixin<HelloPass> {
+public:
+  PreservedAnalyses run(Function &F, FunctionAnalysisManager &AM);
+};
+
+}  // namespace llvm
+
+#endif
+```
+
+Edit `Hello.cpp` as follows:
+
+```cpp
+#include "llvm/Transforms/Utils/Hello.h"
+
+namespace llvm {
+
+PreservedAnalyses HelloPass::run(Function &F, FunctionAnalysisManager &AM) {
+  errs() << F.getName() << "\n";
+  return PreservedAnalyses::all();
+}
+
+}  // namespace llvm
+```
+
+Yes! This is (almost) all you need to write a simple but working pass. Let's break the code into pieces and see how it's working.
+
+To create a pass in C++, you need to write a class that implements the software interface of a pass. Unlike traditional approaches which rely on class inheritance, the new pass manager uses _concepts-based polymorphism_ [^concepts-based-polymorphism]. **As long as the class contains a `run` method that allows it to run on some piece of IR, it is a pass**. No need to inherit the class from some base class and override some virtual functions. In the code above, our `HelloPass` class inherits from the `PassInfoMixin` class, which adds some boilerplate code to the pass. But the most important part is the `run` method that makes `HelloPass` a pass.
+
+[^concepts-based-polymorphism]: For a detailed introduction and discussion about concepts-based polymorphism, please refer to [here](https://gist.github.com/GuillaumeDua/b0f5e3a40ce49468607dd62f7b7809b1).
```

Comment:
IIRC the concept based polymorphism used here is Sean Parent's idea, and there's actually a comment in the code that explains this:
https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/IR/PassManager.h#L27-L33

It would be nice if you can mention it here.

Reply:
Added.